include/nbl/builtin/hlsl/emulated/emulated_float64_t_utils.hlsl

```diff
@@ -0,0 +1,459 @@
+#else
+template<typename device_capabilities = void, bool FastMathIfEmulated = false, bool FlushDenormToZeroIfEmulated = true>
+using portable_float64_t = typename conditional<true, float64_t, emulated_float64_t<FastMathIfEmulated, FlushDenormToZeroIfEmulated> >::type;
+#endif
+
+template<typename EmulatedType, uint32_t N>
+struct emulated_vector {};
+
+template<typename EmulatedType>
+struct emulated_vector<EmulatedType, 2>
+{
+    using type = emulated_vector<EmulatedType, 2>;
+
+    EmulatedType x;
+    EmulatedType y;
+
+    EmulatedType calcComponentSum() NBL_CONST_MEMBER_FUNC
+    {
+        return x + y;
+    }
+
+    NBL_CONSTEXPR_STATIC_INLINE type create(EmulatedType x, EmulatedType y)
+    {
+        type output;
+        output.x = x;
+        output.y = y;
+
+        return output;
+    }
+
+    type operator+(float rhs)
+    {
+        type output;
+        EmulatedType rhsAsEF64 = EmulatedType::create(rhs);
+        output.x = x + rhsAsEF64;
+        output.y = y + rhsAsEF64;
+
+        return output;
+    }
+
+    type operator+(EmulatedType rhs)
+    {
+        type output;
+        output.x = x + rhs;
+        output.y = y + rhs;
+
+        return output;
+    }
+
+    type operator+(type rhs)
+    {
+        type output;
+        output.x = x + rhs.x;
+        output.y = y + rhs.y;
+
+        return output;
+    }
+
+    type operator-(float rhs)
+    {
+        return create(x, y) + (-rhs);
+    }
+
+    type operator-(EmulatedType rhs)
+    {
+        return create(x, y) + (rhs.flipSign());
+    }
+
+    type operator-(type rhs)
+    {
+        rhs.x = rhs.x.flipSign();
+        rhs.y = rhs.y.flipSign();
+        return create(x, y) + rhs;
+    }
+
+    type operator*(float rhs)
+    {
+        type output;
+        EmulatedType rhsAsEF64 = EmulatedType::create(rhs);
+        output.x = x * rhsAsEF64;
+        output.y = y * rhsAsEF64;
+
+        return output;
+    }
+
+    type operator*(EmulatedType rhs)
+    {
+        type output;
+        output.x = x * rhs;
+        output.y = y * rhs;
+
+        return output;
+    }
+
+    type operator*(type rhs)
+    {
+        type output;
+        output.x = x * rhs.x;
+        output.y = y * rhs.y;
+
+        return output;
+    }
+};
+
+template<typename EmulatedType>
+struct emulated_vector<EmulatedType, 3>
+{
+    using type = emulated_vector<EmulatedType, 3>;
+
+    EmulatedType x;
+    EmulatedType y;
+    EmulatedType z;
+
+    EmulatedType calcComponentSum() NBL_CONST_MEMBER_FUNC
+    {
+        return x + y + z;
+    }
+
+    type operator*(NBL_CONST_REF_ARG(type) rhs) NBL_CONST_MEMBER_FUNC
+    {
+        type output;
+        output.x = x * rhs.x;
+        output.y = y * rhs.y;
+        output.z = z * rhs.z;
+
+        return output;
+    }
+};
+
+template<typename EmulatedType>
+struct emulated_vector<EmulatedType, 4>
+{
+    using type = emulated_vector<EmulatedType, 4>;
+
+    EmulatedType x;
+    EmulatedType y;
+    EmulatedType z;
+    EmulatedType w;
+};
```

Comment:
write yourself a `T __array_get(uint32_t i)` and `void __array_set(uint32_t i, T val)` (since HLSL doesn't support `operator[]` returning a reference)

and for each specialization of `emulated_vector_base` only declare the `x,y,z,w` variables, the "pseudo array operators", and a `const static uint32_t Dimension`

then you can write all your operators using CRTP
```cpp
template<typename CRTP>
struct emulated_vector :  CRTP
{
};
```
using `for` loops conditioned on `Dimension`, etc (don't worry about them, they'll get flattened by the compiler)